### ISSUE TITLE
Save xlsx file from gsheets alters file

### ISSUE THREAD

**Issue by nicosafull** [xlnt-community/xlnt#5]
Hello,

I am experiencing an issue with the xlnt library when trying to read spreadsheets that have been saved from Google Sheets. When I open a spreadsheet with MS Excel and save it as an xlsx file, the xlnt library is able to parse and save the file correctly. However, if I open the same file with Google Sheets and save it again (keeping the xlsx format), the xlnt library no longer "understands" the empty/null cells.

Here is a code snippet that demonstrates the issue:

```cpp
wb.load(DatFileNames.As_Ansi());

if (!wb.contains("Functions"))
    return FALSE;

//-- Find the sheet.
auto ws = wb.sheet_by_title("Functions");

//-- Get the number of records excluding header.
auto rows = ws.rows();
m_NumFunctions = rows.length() - 1;
```

In this case, `m_NumFunctions` returns 53 when the file is saved with MS Excel, which is the correct number of rows in the spreadsheet. However, when the file is saved with Google Sheets then `m_NumFunctions` returns 999.

When I iterate through the rows, I encounter another issue. Here is the code snippet:

```cpp
for (size_t i = 0; i < m_NumFunctions; i++)
{
    size_t rowIndex = i + 2; // 2 as offset because we must consider Excel is not zero based and then because we want to skip first row as it's the header
    CommandID = rows.cell(L_String("A%d", rowIndex).As_Ansi()).value<int>();
}
```

In this case, `CommandID` on the actual last row is 53, which is correct. However, on the following iteration (that only happens when file is saved on Google Sheets as on MS Excel it just finishes because it read the legth properly) `CommandID` returns "0" (zero), even though there is no actual zero in the spreadsheet. I have seen other users struggling with similar issues when using Google Sheets, as it seems to be defaulting a zero on cells that it cannot properly set.

I am not sure how to make the xlnt library "aware" that "0" is not a legitimate value but an actual end-of-line or terminator. I would appreciate any help or suggestions on how to resolve this issue.

Thank you.

**Comment by m7913d** [comment]
Can you also share a (simplified) example of a file that reproduces this issue?

**Comment by m7913d** [comment]
Concerning the `0` being returned: if the cell is empty, `0` is indeed returned. If you want to determine the difference between empty and `0`, you may call the `has_value` function or call `data_type` to check the data type of the cell.

**Comment by nicosafull** [comment]
thing is that if you create an Excel file, add 3 rows and save it with MS Excel, then xlnt will work and dimension will be ok, now open that same file with gsheets and just save (Export -> Save as XLSX) and now dimension will be 999, and from the 3rd row, all cell values will be/contain "0" though they don't hold anything

**Comment by nicosafull** [comment]
my problem is that dimension is unreliable if you save a file through gsheets, as it will return 999 so I can't tell the real end.. dimension should return the same regardless if fale is saved from gsheets or MS Excel, there is something gsheet does.

That's the test, try any xlsx file in which xlnt will return a fixed version like, let's say 4, then open that very same file on gsheets, re-save it from gsheets as a different copy (so you don't override MS Excel one) and now dimension will return 999.

This happens only from gsheets, libreoffice and openoffice won't do this

**Comment by nicosafull** [comment]
> Concerning the `0` being returned: if the cell is empty, `0` is indeed returned. If you want to determine the difference between empty and `0`, you may call the `has_value` function or call `data_type` to check the data type of the cell.

my problem is that dimension is unreliable if you save a file through gsheets, as it will return 999 so I can't tell the real end.. dimension should return the same regardless if fale is saved from gsheets or MS Excel, there is something gsheet does.

That's the test, try any xlsx file in which xlnt will return a fixed version like, let's say 4, then open that very same file on gsheets, re-save it from gsheets as a different copy (so you don't override MS Excel one) and now dimension will return 999.

This happens only from gsheets, libreoffice and openoffice won't do this

**Comment by m7913d** [comment]
I'm afraid this is an Google Sheets "issue". Inspecting the internals of the XLSX file generated by Google Sheet, google sheets actually saves 1000 rows, many rows containing no cells.
If you want to know the number of non-empty cells, you should iterate over the range returned by `rows` and check for the last non empty cell.

**Comment by nicosafull** [comment]
so `worksheet.rows();` with the `skip_null` isn't an option, as I know technically they aren't null?

**Comment by doomlaur** [comment]
@nicosafull The Pull Request made by @m7913d should fix your issue. I just merged it to the master branch. Please let us know if you have any further related issues.

**Comment by doomlaur** [comment]
@nicosafull At the risk of being slightly off-topic, I noticed the following line in your code:
```cpp
wb.load(DatFileNames.As_Ansi());
```
I noticed that you are calling `As_Ansi()`, which made me worry about the correctness of file path encodings in your code. You did not mention which operating system you are using XLNT on, but I just want to shortly mention the encodings that XLNT uses for file paths:
| OS | String type | Encoding |
| ----- | ----- | ----- |
| Windows / Visual Studio | `std::string` | UTF-8 |
| Windows / Visual Studio | `std::wstring` | UTF-16 |
| Non-Windows / non Visual Studio | `std::string` | Locale-encoded (usually UTF-8, but not guaranteed) |

Note: `std::wstring` is only available in XLNT under Windows with Visual Studio.

In other words, when using file paths, you will want to do one of the following to ensure that paths are correctly handled with all Unicode characters (as much as possible, at least):

1. Use `std::wstring` (with UTF-16) on Windows and `std::string` (locale-encoded, but usually UTF-8) on other operating systems.
2. Use `std::string` on Windows, but you must ensure UTF-8 encoding (!), and also `std::string` on non-Windows (with locale-encoded strings).

**Comment by m7913d** [comment]
@nicosafull Note that `ws.rows().length()` does also not return the largest row number in general, but the number of rows between the first and last non-empty row (including first and last row). So, in case of empty starting rows, the result is different from what you want. The function `highest_row` is probably what you want to use.

**Comment by nicosafull** [comment]
@doomlaur I'm using 16-bit UNICODE characters

@m7913d what would be the proper code then? `worksheet.calculate_dimension()` is still returning 999 on both, width and length. But if I do `worksheet.columns().length()` and  `worksheet.rows().length()` it works ok

**Comment by m7913d** [comment]
The function you want is `worksheet::highest_row()`: this function return the highest row containing non-empty cells.

The same result can be obtained by `worksheet::calculate_dimension(true, true).bottom_right().row()`, but this is just less efficient if you are only interested in the highest row. By default, this function behavior is not changed by the fix, but by passing true as the second argument, you should get the desired result (i.e. ignore rows for which only properties are set by Google Sheets).
As stated before `worksheet.rows().length()` returns the "wrong" result if the top row(s) are empty, i.e. it returns (highest_row - lowest_row + 1), but this probably doesn't matter in your use case.

`worksheet.columns().length()` is definitely wrong as it returns a number of columns, not rows.

`worksheet.rows()` internally calls `worksheet::calculate_dimension(true, true)`. So, you can also obtain the desired result by `worksheet.rows().reference().bottom_right().row()`.

I hope this makes it clear.

**Comment by nicosafull** [comment]
yes, thank you
